src/main.c

```diff
@@ -27,59 +27,95 @@ static int g_sequence = 0;
 
 #define SEQUENCE_MASK (0xffffffff ^ (0xffffffff << SEQUENCE_BITS))
 
-static long get_timestamp() {
+struct conf
+{
+    long snowflake_epoc;
+    int node_id_bits;
+    int datacenter_id_bits;
+    int sequence_bits;
+    int node_id_shift;
+    int datacenter_id_shift;
+    int timestamp_shift;
+    int sequence_mask;
+} conf;
+
+static long get_timestamp()
+{
     struct timeval tv;
     gettimeofday(&tv, NULL);
 
     return tv.tv_sec * 1000 + tv.tv_usec / 1000;
 }
 
-static long get_til_next_millis(long last_timestamp) {
+static long get_til_next_millis(long last_timestamp)
+{
     long ts = get_timestamp();
-    while (ts < last_timestamp) {
+    while (ts < last_timestamp)
+    {
         ts = get_timestamp();
     }
 
     return ts;
 }
 
-static int luasnowflake_init(lua_State *L) {
+static int luasnowflake_init(lua_State *L)
+{
+    conf.snowflake_epoc = luaL_optlong(L, 3, SNOWFLAKE_EPOC);
+    conf.node_id_bits = luaL_optint(L, 4, NODE_ID_BITS);
+    conf.datacenter_id_bits = luaL_optint(L, 5, DATACENTER_ID_BITS);
+    conf.sequence_bits = luaL_optint(L, 6, SEQUENCE_BITS);
+
+    if ((conf.node_id_bits + conf.datacenter_id_bits + conf.sequence_bits) > 32)
```

Comment:
Should we also count the snowflake_epoc?
We should also ensure all fields are >= 1